### PR TITLE
Crosslink helpers

### LINKED ISSUE #10: update logic for crosslink notaries
# Issue
The logic for crosslink notaries in the current implementation is lacking. It assumes we will be able to go through every shard every epoch. This is not actually the case because there is to be a fixed number of notaries (`NOTARIES_PER_CROSSLINK`) per crosslink, but a constant number of shards and variable number of validators. 

# Proposed Implementation
- Add global var `NOTARIES_PER_CROSSLINK`
- add helper function `get_crosslink_shards` to get the shards that should be crosslinked this epoch
- add helper function `get_crosslink_notaries` to get the validators that should notarize the given shard for this epoch.
- remove helper function `get_shard_attesters` (superseded with `get_crosslink_notaries`)
- utilize these helper functions to limit the work to be done by validators on shards per epoch. _do not_ accept crosslinks for shards not in `get_crosslink_shards`

Implementations of the helper functions are approximately the following:
```python
def get_crosslink_shards(crystallized_state):
    start_from = crystallized_state.next_shard
    count = len(crystallized_state.active_validators) // NOTARIES_PER_CROSSLINK
    if start_from + count <= SHARD_COUNT:
        return list(range(s, s+count))
    else:
        return list(range(start_from, SHARD_COUNT)) + list(range(start_from + count - SHARD_COUNT))
```

```python
def get_crosslink_notaries(crystallized_state, shard_id):
    crosslink_shards = get_crosslink_shards(crystallized_state)
    if shard_id not in crosslink_shards:
        return None
    all = len(crystallized_state.current_shuffling)
    start = all * crosslink_shards.index(shard_id) // len(crosslink_shards)
    end = all * (crosslink_shards.index(shard_id)+1) // len(crosslink_shards)
    return crystallized_state.current_shuffling[start: end]
```

### PR DESCRIPTION
**Pull request by hwwhww**
# What's the problem?
Fix #10

# What did I do:
* Copied the code from the spec.
* Added some edge cases handling and refactor it.

### CODE REVIEW

File: tests/state/test_helpers.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+
+@pytest.mark.parametrize(
+    'active_validators_count, expected',
+    (
+        (1000, 10),
+        (10000, testing_helpers_config['shard_count'])
+    )
+)
+def test_get_crosslink_shards_count(active_validators_count, expected):
```
Comment (djrtwo) [inline tests/state/test_helpers.py]: I prefer to use config fixtures like this. By overriding `shard_count` and `notaries_per_crosslink` in the `parametrize` it actually ripples through the `config` fixture that you get.
```python
@pytest.mark.parametrize(
    'active_validators_count,shard_count,notaries_per_crosslink,expected',
    (
        (1000, 20, 100, 10),
        (10000, 20, 100, 20),
        (50, 100, 3, 16),
    )
)
def test_get_crosslink_shards_count(active_validators_count,
                                    shard_count,
                                    notaries_per_crosslink,
                                    config,
                                    expected):
    crosslink_shards_count = get_crosslink_shards_count(
        active_validators_count,
        config=config,
    )
    assert crosslink_shards_count == expected
```
(this hinges upon defining a `notaries_per_crosslink` fixture and having it feed into the `config` fixture like the rest of the config vars)

I find it a little bit more readable, and it also gives you the ability to easily try more varied combinations of configurations in your test. Notice the `(50, 100, 3, 16)` test that changes up the config.

File: tests/state/test_helpers.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        ),
+    ),
+    ids=[
+        'next_shard=0, 1000 validators',
+        'next_shard=10, 1000 validators',
+        'next_shard=19, 1000 validators',
+        'next_shard=10, 10001 validators',
+    ],
+)
+def test_get_crosslink_shards_and_get_crosslink_notaries(
```
Comment (djrtwo) [inline tests/state/test_helpers.py]: If you follow the method I outlined above, it would be `shard_count`, `notaries_per_crosslink`, and `num_validators` as parameters and dynamically build the `active_validators` in-test. Would allow you to be more explicit and varied in the config you are testing

File: tests/state/test_helpers.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+}
+testing_helpers_config = generate_config(
+    shard_count=20,
+    notaries_per_crosslink=100,
+)
+
+
+def create_crystallized_state(
+        genesis_crystallized_state,
+        init_shuffling_seed, next_shard,
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply tests/state/test_helpers.py]: oops, I missed it!

File: beacon_chain/state/helpers.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+def get_crosslink_notaries(
+        crystallized_state,
+        shard_id,
+        config=DEFAULT_CONFIG):
+    """Returns a list of notaries that will notaries `shard_id`
+    """
+    crosslink_shards = get_crosslink_shards(crystallized_state, config=config)
+    if shard_id not in crosslink_shards:
+        return None
+    all = len(crystallized_state.current_shuffling)
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply beacon_chain/state/helpers.py]: 1. Agreed with the renaming!
2. I found that we use `len()` a lot. We can also add some properties for `CrystallizedState` and `ActiveState`, e.g.,
    ```python
    @property
    def num_active_validators(self):
        return len(self.active_validators)
    ```

[the remark this answers is not in the document]
Comment (djrtwo) [reply beacon_chain/state/helpers.py]: Absolutely

**Comment by hwwhww** [comment]
@djrtwo Thanks for the feedback! The tests look much cleaner with your suggestions. 🙂 
I added "utilizing these helper functions" part in this PR, and I will add `num_` proproties in another PR.

**Comment by djrtwo** [comment]
Looks good! I fixed a couple of typos in comments.
The tests cleaned up nicely. Things are starting to feel better in here 😄
